### DISCUSSION TITLE
Allow Hosting Laravel Applications Locally on the Network with Herd

### DISCUSSION THREAD

**Discussion by Murtaza1904** [beyondcode/herd-community#1197] [General]
### Body Content:  

I have been using Laravel Herd for local development and have found it to be an excellent tool for running Laravel applications. However, I have encountered a limitation when attempting to host my Laravel application locally on the network.  

In a traditional setup using the `php artisan serve` command, we can specify the host IP to serve the application on the local network, allowing other devices on the same network to access the application. For instance:  

```bash  
php artisan serve --host=192.168.x.x --port=8000  
```  

This feature is especially useful in scenarios like:  
- Testing applications on physical devices (e.g., mobile or tablet).  
- Collaborating with team members or stakeholders over a shared network.  
- Quickly showcasing the application without deploying it to a server.  

Currently, Laravel Herd does not provide an option to configure the host IP or expose the application to the local network. It would be highly beneficial if Herd could support this feature, either by extending the existing functionality or by providing an additional configuration option in its setup.  

#### Suggested Solution:  
1. Add support for configuring the host and port through Herd's command-line interface or settings.  
2. Allow binding to specific IP addresses to expose the application on the local network.  
3. Provide clear documentation on how to achieve this once implemented.  

#### Benefits:  
- Enhanced flexibility for developers.  
- A more streamlined workflow for testing on multiple devices and collaborating in real-time.  
- Reduces the need for additional tools or manual workarounds for local network hosting.  

I hope this feature can be considered for a future release, as it would make Laravel Herd an even more versatile tool for local development. Thank you for your efforts in maintaining this project!  

Looking forward to your feedback on this suggestion.

**Comment by mihas7f** [comment]
I recently purchased Laravel Herd Pro, and I'm quite disappointed to find out that this functionality is not available out of the box. I hope the team considers adding this as a default feature in the future.

**Comment by zedhmoud** [comment]
i have the same issue , so is their another way to resolve it ?

**Comment by sschlein** [comment]
The difficulty with this request is that Herd serves applications via their local `.test`domains and that this isn't available anymore when you access your device via the IP, so there is no way of Herd knowing which site to serve. You can use Expose (https://expose.dev) for exposing your local site and using the Expose link (the free version should be just fine)

**Comment by pr0ph3tx** [comment]
Until Herd supports LAN binding natively, here’s a simple workaround to expose specific projects on specific ports (e.g., http://192.168.1.70:5001 → project-x.test, :5002 → project-y.test). This is Windows-focused, but the idea is the same elsewhere.

1. Open Herd’s Nginx folder (Path is usually %USERPROFILE%\.config\herd\config\nginx)
3. Create a file (e.g. mysites.conf) make sure it’s **.conf**, not .conf.txt
4. Add one server block per site/port (duplicate & edit as needed)

```
server {
    listen 0.0.0.0:5001;  #<- YOUR CUSTOM PORT HERE
    server_name _;

    # Send the request to Herd's HTTP on localhost,
    # but tell upstream we want project-x.test
    location / {
        proxy_set_header Host yousite.test;        # <- YOUR HERD SITE HERE
        proxy_set_header X-Forwarded-Host $http_host;# keep client-visible host (IP:port)
        proxy_set_header X-Forwarded-Proto $scheme;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Port $server_port;
        proxy_pass http://127.0.0.1:80;
    }

    access_log off;
}
```

Repeat for 5002 → anothersite.test, etc.

5. Open %USERPROFILE%\.config\herd\config\nginx\nginx.conf
6. Scroll all the way to the bottom and right after "include herd.conf" add "include mysites.conf;"
7. Restart all services

Now you can host any site you have at any custom port. (providing you have the port available)
For others to connect to your sites, simply look for your internal IP address, usually 192.168.X.X and then append your custom port.

Happy development!

**Reply by iainco** [reply]
Thank you! Works like a charm. On MacOS the conf files are here: `/Users/USERNAME/Library/Application Support/Herd/config/nginx`

**Comment by tommy66374** [comment]
My solution is based on this https://laravel.com/docs/12.x/valet#sharing-sites-on-your-local-network

In my Windows 11 computer i have a Herd secure site for example `https://lan-only.test`

backup
`C:\Users\<username>\.config\herd\config\valet\Nginx\lan-only.test.conf`
as
`C:\Users\<username>\.config\herd\config\valet\Nginx\lan-only.test - original backup.conf`

edit
`C:\Users\<username>\.config\herd\config\valet\Nginx\lan-only.test.conf`

```diff
server {
-   listen 127.0.0.1:80;
+   listen 80;
    #listen 127.0.0.1:80; # valet loopback
    server_name lan-only.test www.lan-only.test *.lan-only.test;
    return 301 https://$host$request_uri;
}

server {
-   listen 127.0.0.1:443 ssl;
+   listen 443 ssl;
    #listen VALET_LOOPBACK:443 ssl; # valet loopback
    server_name lan-only.test www.lan-only.test *.lan-only.test;
    root /;
    charset utf-8;
    client_max_body_size 512M;
    http2 on;

+   allow 127.0.0.1;
+   allow 192.168.1.33; # <--- THIS IS ANOTHER LOCAL DEVICE IP
+   deny all;

    location ~* /<hash-string>/([A-Z]+:)(.*) {
        internal;
        alias $1;
        try_files $2 $2/;
    }

    # OTHER CODE BELOW
```

Save this file and go to Herd Dashboard restart all services and also in this computer firewall allow port 80, 443 come to `C:\Program Files\Herd\resources\app.asar.unpacked\resources\bin\nginx\nginx.exe`

If an IP not in the above `allow IP` will get 403 web server error, and will add an error into `C:\Users\<username>\.config\herd\Log\nginx-error.log`

You shouldn't open the WAN port 80, 443 in your router.
